rust-src/concordium_base/src/sigma_protocols/ps_sig_known.rs

```diff
@@ -0,0 +1,873 @@
+    fn serial<B: Buffer>(&self, out: &mut B) {
+        match self {
+            Self::EqualToCommitment(m, cmm) => {
+                out.put(&0u8);
+                out.put(m);
+                out.put(cmm);
+            }
+            Self::Public => {
+                out.put(&1u8);
+            }
+            Self::Known(m) => {
+                out.put(&2u8);
+                out.put(m);
+            }
+        }
+    }
+}
+
+impl<C: Curve> Deserial for PsSigSecretMsg<C> {
+    fn deserial<R: ReadBytesExt>(source: &mut R) -> ParseResult<Self> {
+        let tag: u8 = source.get()?;
+        Ok(match tag {
+            0 => {
+                let m = source.get()?;
+                let cmm = source.get()?;
+                Self::EqualToCommitment(m, cmm)
+            }
+            1 => Self::Public,
+            2 => {
+                let m = source.get()?;
+                Self::Known(m)
+            }
+            _ => anyhow::bail!("unsupported SecretMsg item type: {}", tag),
+        })
+    }
+}
+
+/// Witness used in proof, maps to the "statement" $y$ under $\varphi$
+pub struct PsSigSecret<P: Pairing, C: Curve<Scalar = P::ScalarField>> {
+    /// Secret $r'$ value
+    pub r_prime: Secret<P::ScalarField>,
+    /// Secret value for each message part
+    pub msgs: Vec<PsSigSecretMsg<C>>,
+}
+
+/// Response in the protocol
+type ResponseMsg<C> = PsSigSecretMsg<C>;
+
+/// Response in sigma protocol
+#[derive(Clone, Debug, Serialize)]
+pub struct Response<P: Pairing, C: Curve<Scalar = P::ScalarField>> {
+    /// The response corresponding to $r'$
+    resp_r_prime: P::ScalarField,
+    /// The response corresponding to each part of the message
+    #[size_length = 4]
+    resp_msgs: Vec<ResponseMsg<C>>,
+}
+
+impl<P: Pairing, C: Curve<Scalar = P::ScalarField>> SigmaProtocol for PsSigKnown<P, C> {
+    type CommitMessage = (P::TargetField, Vec<Commitment<C>>);
+    type ProtocolChallenge = C::Scalar;
+    type ProverState = PsSigState<P, C>;
+    type Response = Response<P, C>;
+    type SecretData = PsSigSecret<P, C>;
+
+    #[inline]
+    fn public(&self, ro: &mut RandomOracle) {
+        ro.add_bytes(b"PsSigKnown");
+        ro.append_message(b"blinded_sig", &self.blinded_sig);
+        ro.extend_from(b"messages", self.msgs.iter());
+        ro.append_message(b"ps_pub_key", &self.ps_pub_key);
+        ro.append_message(b"comm_key", &self.cmm_key)
+    }
+
+    #[inline]
+    fn get_challenge(
+        &self,
+        challenge: &crate::random_oracle::Challenge,
+    ) -> Self::ProtocolChallenge {
+        C::scalar_from_bytes(challenge)
+    }
+
+    /// Compute commit secrets $\alpha$ and their image $a = \varphi(\alpha)$ under $\varphi$ (see module [`self`] for definition of $\varphi$).
+    #[inline]
+    fn compute_commit_message<R: Rng>(
+        &self,
+        csprng: &mut R,
+    ) -> Option<(Self::CommitMessage, Self::ProverState)> {
+        let g_tilde = self.ps_pub_key.g_tilda;
```

Comment:
In the current bluepaper, this generator is know as `g_2` (`g_BLS,2` to be precise). Should we rename it here as well?